questions.md

```diff
@@ -4,26 +4,125 @@
+
 3. In order to be a "good citizen", you need to have some idea of much RAM your job requires.  In particular, you need
    to know the "peak" (i.e., maximum) RAM required at any point during execution.  Show an example of the shell command
-   that you would use on a Linux machine to measure run time and peak ram usage of an arbitrary command, where the time/peak RAM values are written to a file.
+   that you would use on a Linux machine to measure run time and peak ram usage of an arbitrary command, where the time/peak        RAM values are written to a file.
+
+Could run your application/command on a small set of input data/subset of input data and scale up from there, prefixed by:
+
+```
+/usr/bin/time -v commandhere 2>&1 outfile.txt
+```
+
+Based on the values given and the initial size of your subsetted input file vs the total size of the original file you could estimate the time and RAM required for the actual job.
+
+   
 4. What are the units of your answer for number 3?
+
+User and system time is reported in seconds
+
+Elapsed wall clock time is hour:min:sec
+
+Maximum resident set size values are reported in kbytes
+
 5. What are the bash commands for the following operations:
 
     * Checking that a file exists
+    
+    ```
+if [ -e txt.md ]
+then
+    echo "Exists"
+else
+    echo "Does not exist"
+fi
+```
+    
     * Checking that a file exists and is not empty
+    
+    ```
+if [ -e test.md ]
+then
+  if [ -s test.md ]
+  then
+   echo "Exists and is full"
+   else
+    echo "Exists but is empty"
+fi
+else
+  echo "Does not exist"
+fi
+```
 
 6. How would you use the commands from your answer to 5 to write a work flow for HPC that only runs a job if the
    expected output file is **not** present.
 
+```
+if [ -e test.txt ]
+then
+   echo "test.txt is here"
+   else
+   executecommandhere.py
+fi
+```
+
 ## Trickier questions
 
 7. Would your answer to number 3 work on Apple's OS X operating system?  If no, do you have any idea why not? 
 
+No, the exact answer provided in number 3 does not work for mac. It only gives the time elapsed in seconds but not the additional information about the memory used/size of written files or peak RAM usage. The verbose option is not available in mac. As an alternative, one could install gnu-time and use the gtime command to get a similar verbose output as question 3.
+
 8. Most of the HPC nodes have 512Gb (gigabytes) of RAM. Let's say you have a job that will require **no more** than 24Gb
    of RAM.  How would you request resources so that you can run more than one job on a node **and** not cause nodes to
    crash?  Show an example of a skeleton HPC script as part of your answer.  **Knowing how to do this is super important
    and will save you loads of frustration and prevent you from taking out your colleagues' jobs on the cluster,
    preventing you from getting nasty emails from Harry!!!!!!!!!!!**
+
+```
+#!/bin/bash
```

Comment:
Hi, this is correct but I would like to add some caveats.

By default and as a common courtesy it is best to take the total amount of RAM on a system, subtract a small overhead (our admins have mostly taken care of this so we don't really need to worry about this value much) and then divide it by the total number of cores. if your job will be using 256GB of RAM on a 512GB RAM node it is best to also use the same portion of cores for your job, even though your job will only run on 1 core. Technically, our grid should have the option to limit the amount of maximum memory requested based on number of cores requested. i.e on a 64 core 512GB node the user should allocate a maximum of 8GB per core. if the single core job were to use 16GB then it would require that 2 cores be allocated thereby avoiding issues caused by the over consumption of RAM by a job.